src/topology/ping_pong.rs

```diff
@@ -0,0 +1,865 @@
+            }
+            Self::Finish { prep_msg } => {
+                2u8.encode(bytes);
+                encode_u32_items(bytes, &(), prep_msg);
+            }
+        }
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        match self {
+            Self::Initialize { prep_share } => Some(1 + 4 + prep_share.len()),
+            Self::Continue {
+                prep_msg,
+                prep_share,
+            } => Some(1 + 4 + prep_msg.len() + 4 + prep_share.len()),
+            Self::Finish { prep_msg } => Some(1 + 4 + prep_msg.len()),
+        }
+    }
+}
+
+impl Decode for Message {
+    fn decode(bytes: &mut std::io::Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let message_type = u8::decode(bytes)?;
+        Ok(match message_type {
+            0 => {
+                let prep_share = decode_u32_items(&(), bytes)?;
+                Self::Initialize { prep_share }
+            }
+            1 => {
+                let prep_msg = decode_u32_items(&(), bytes)?;
+                let prep_share = decode_u32_items(&(), bytes)?;
+                Self::Continue {
+                    prep_msg,
+                    prep_share,
+                }
+            }
+            2 => {
+                let prep_msg = decode_u32_items(&(), bytes)?;
+                Self::Finish { prep_msg }
+            }
+            _ => return Err(CodecError::UnexpectedValue),
+        })
+    }
+}
+
+/// A transition in the pong-pong topology. This represents the `ping_pong_transition` function
+/// defined in [VDAF].
+///
+/// # Discussion
+///
+/// The obvious implementation would of `ping_pong_transition` would be a method on trait
+/// [`PingPongTopology`] that returns `(State, Message)`, and then `ContinuedValue::WithMessage`
+/// would contain those values. But then DAP implementations would have to store relatively large
+/// VDAF prepare shares between rounds of input preparation.
+///
+/// Instead, this structure stores just the previous round's prepare state and the current round's
+/// preprocessed prepare message. Their encoding is much smaller than the `(State, Message)` tuple,
+/// which can always be recomputed with [`Self::evaluate`].
+///
+/// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+#[derive(Clone, Debug, Eq)]
+pub struct Transition<
+    const VERIFY_KEY_SIZE: usize,
+    const NONCE_SIZE: usize,
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+> {
+    previous_prepare_state: A::PrepareState,
+    current_prepare_message: A::PrepareMessage,
+}
+
+impl<
+        const VERIFY_KEY_SIZE: usize,
+        const NONCE_SIZE: usize,
+        A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+    > Transition<VERIFY_KEY_SIZE, NONCE_SIZE, A>
+{
+    /// Evaluate this transition to obtain a new [`State`] and a [`Message`] which should be
+    /// transmitted to the peer.
+    #[allow(clippy::type_complexity)]
+    pub fn evaluate(
+        &self,
```

Comment:
I'd like to do that, but that ends up being impractical. What I observed in Janus is that I need to have a live reference to both the `Transition` and the `Message` returned from `evaluate` so I can store the former in a database and transmit the latter to the peer. So if we made `evaluate` consume `self`, we wouldn't be getting rid of clones, we'd just be pushing them into the caller.

Reply:
I think this is a good idea; there is currently some API impedance-mismatch in Janus' usage of this method that means we need to call `evaluate` and then still keep the `Transition` around, but I think we need to resolve that to avoid calling `evaluate` multiple times on the same `Transition`.

Reply:
There's no harm in calling `evaluate` multiple times, but I suppose I, too, like the idea of consuming `self` here. I'll tinker with this some more to see if I can make this work out.

Reply:
I do not think we can avoid calling `evaluate` twice, at least not without forcing a helper to hang on to more data when it's in the waiting state. The point of `Transition` is to allow users of `prio` to choose whether to spend more compute (by calling `evaluate` twice) or more storage space (by calling `evaluate` once and then having to store relatively large prepare message shares).

As for making this method consume `self`: I don't think it's worth it. Doing that forces a lot of awkward clone gymnastics into the caller instead of doing them here. In the case where the `Transition` is dropped right away, I suspect the compiler will optimize the clones away anyway. Finally, there's no harm in calling `evaluate` more than once on the same `Transition`: you'll just get the exact same result back. So I'm not certain we're improving the ergonomics or safety of the API by doing that.

Reply:
For Prio3 the PrepState includes the output share (IIRC). For something like Prio3SumVec, this may be quite large, perhaps a megabyte or so. I think it's worth figuring out what the compiler does and if we can easily avoid a copy here.

Reply:
Correction: It includes the measurement share, which is short for the Helper (just a seed) and long for the Leader. For the Leader we can probably optimize this a bit more by only storing the truncated measurement share, but that still doesn't help much for Prio3SumVec.